### DISCUSSION TITLE
Additional Consumption Reports

### DISCUSSION THREAD

**Discussion by coolbung** [Sunbird-Ed/Community#238] [Enhancement request]
We're needing additional reports related to consumption. Sharing below the list of reports and the columns needed in each report. 

**Course Progress Report**
Course ID | Course Name | Course Created By | Batch | Batch Start Date | Users Enrolled | Users Completed | Avg Time Spent

**Content Progress Report**
Content ID | Content Name | Course ID | Course Name | Content Created By | Batch | Batch Start Date | Completions | In Progress | Avg Time Spent

**Learner Report**
User ID | User Name | User Email | User Phone | Last Login | Enrolled Courses | Completed Courses | Incomplete Courses | Total Time Spent

**Learner Courses Report**
User ID | User Name | User Email | User Phone | Course Name | Batch | Enrollment Date | Completion Status | Completion Percentage | Completion Date | Time Spent | Certificate 

## Approach
1. Create list and summary APIs based on the `user_courses` and `user_content_consumption` cassandra tables
2. A new private user search API that returns un-encrypted user data based on the consent provided. If the consent is turned off for the tenant, the data will be always returned unencrypted. The API should be able to return multiple users in a single call
3. A new course reports service to be created that orchestrates the data by calling above services. This service will have APIs for each of the reports above 

cc @santhoshv

**Comment by Krishnaj20** [comment]
@coolbung - thanks for initiating this thread. As next steps, if you can come up with the implementation/design approach doc we can open it up for feedback from the working group. 
@reshmi-nair

**Reply by coolbung** [reply]
@reshmi-nair could you share the details of the available services related to aggregate / consumption data ?

**Reply by Krishnaj20** [reply]
@reshmi-nair - following up on this thread. Appreciate if you can share these details so that @coolbung can come up with the initial design doc.

@rhwarrier @SanthoshVasabhaktula @sharathkashyap

**Reply by reshmi-nair** [reply]
@coolbung Please refer this link to check the collection summary and group agregate APIs 
[Batch-Service API Documentation](https://lern.sunbird.org/learn/product-and-developer-guide/batch-service/api-documentation)

**Reply by Krishnaj20** [reply]
@coolbung - let us know if there are any additional details needed.

**Reply by coolbung** [reply]
@reshmi-nair -- this API `/api/data/v1/group/activity/agg` is for aggregate consumption data against a group. 

Are there any APIs that let us fetch data from the `user_courses` `user_content_consumption` and `user_enrolment` tables ?

**Reply by Krishnaj20** [reply]
@reshmi-nair - can you look into this query?

**Reply by reshmi-nair** [reply]
@coolbung We don't have the API to fetch consumption and enrolment details as of now. We have on demand exhaust reports currently fetching this data. [Exhaust Reports](https://lern.sunbird.org/learn/product-and-developer-guide/batch-service/reports)

**Reply by Krishnaj20** [reply]
tagging @SanthoshVasabhaktula to this thread.

**Reply by SanthoshVasabhaktula** [reply]
@maheshkumargangula - Can you provide the requested details?

**Reply by coolbung** [reply]
Thanks. Do these exhausts run directly on the database ? 

Or is Druid & data pipeline a pre-requisite for these exhaust reports ?

**Reply by SanthoshVasabhaktula** [reply]
@coolbung - While Mahesh is collating the details, do note that all the data might not be at one location or served through one API, but all the required data exists in transactional or analytics or search databases. Your recommended approach of creating a collection summary or report service should be the way forward to aggregate the information from various sources.

**Reply by maheshkumargangula** [reply]
@coolbung - The Exhaust Reports fetch the data directly from Databases (Cassandra for Course and Content Progress).

The existing Exhaust reports doesn't have the required information directly. Please use the below details to design and implement them.

**Course Progress Report:**
**Option 1:** Collection Summary Snapshot Report injected to Druid which has the enrolment and completion counts for all the course batches. 
**Option 2:** There is an AUDIT event for Course Batch enrolment by a user and completion. Please use Druid APIs to access this data and cache it with configurable TTL to reuse. It requires design and enhancements to compute the avg. time spent on a course/batch.
**Option 3:** Create a Spark Script to compute the enrolments and completion using `user_enrolments` table from Cassandra and store it. Run it as a batch script (may be weekly).

**Content Progress Report:**
**Option 1:** There is an AUDIT event for each completion of content consumption (with the required context details - course, batch etc,.) . Here also, use the Druid API to access the required metrics.
**Option 2:** Create a Spark Script to compute the content consumption completion and in progress counts. Use the `content_consumption` Cassandra table to access the data. Please join with `user_enrolments` and `course_batch` table to access `content_consumption` table easily.

**Learner Report:**
Please use the AUDIT events (Course Batch - enrolment and completion) to get the required metrics from Druid or use the `user_enrolemnts` table to compute the required metrics.

**Learner Courses Report:**
[Progress Exhaust](https://lern.sunbird.org/learn/product-and-developer-guide/batch-service/reports#progress-exhaust) report has all the details. The report is generated for each course batch (on request bases).

1. Please use the user cache (redis) to join the user metadata along with the report.
2. To compute the `Time spent` (avg or total) it requires  design and enhancements to the existing implementation.

In the above description, the dimensions for each report are not clear. Please mention below details.
 - the report execution frequency 
 - is it per course/batch or for all course/batches.
 - is it for all users and courses or users of a specific course/batch (Learner Report)

Cc: @SanthoshVasabhaktula @Krishnaj20

**Reply by manjirisulekavadi** [reply]
Also, are we considering all the consumption reports with respect to the contributor type and projects (coming from coKreate) - Individual contributor from the sourcing organization, contributing organization and the individual contributors their within, individual nominated contributors themselves?

**Reply by Krishnaj20** [reply]
@manjirisulekavadi - not sure I understand your query. It will help to know what kind of reports are you looking from coKreat, and are these requirements met/un-met by the existing coKreat data products.

**Reply by manjirisulekavadi** [reply]
@Krishnaj20 As per coKreate documentation for roles
1. Admin of Contributor Organization can monitor the published assets from its consumption
2. Individual Contributor (from contributing/sourcing organization and/or independent??) can monitor usage of published assets 

My question is do we need additional consumption analysis as feature to see the consumption pattern for reach/usage for

Different projects by its contributing/sourcing organizations and individual contributor from  contributing/sourcing organizations or totally independent. (The onboarded organization might be interested in analysis content consumption to track down the quality and/or reach issues.

**Comment by manjirisulekavadi** [comment]
Can a ready framework be used of cQube if Emission data is managed from SB inQiery that handles ED data?

**Reply by Krishnaj20** [reply]
@manjirisulekavadi - I spoke with Ashwin on the queries reported by you in this thread. This thread is related to the consumption reports that @coolbung is exploring to enable for the ED SAAS adopters. 

Request you to create separate individual discussion threads for creation related reports and cQube related questions so that they can be addressed in those threads, and will benefit users who have similar queries in the future. 
@coolbung
